includes/class-llms-form-field-data.php

```diff
@@ -0,0 +1,332 @@
+		 *
+		 * @since [version]
+		 *
+		 * @param string[] $stores An array of data store names.
+		 */
+		return apply_filters(
+			'llms_form_field_data_stores',
+			array(
+				'users',
+				'usermeta',
+			)
+		);
+
+	}
+
+	/**
+	 * Sanitize data before an insertion
+	 *
+	 * @since [version]
+	 *
+	 * @param string $key   Column name.
+	 * @param mixed  $value Data to be inserted.
+	 * @return mixed
+	 */
+	public static function sanitize( $key, $value ) {
+
+		if ( 'protected' === $key ) {
+			return llms_parse_bool( $value ) ? 1 : 0;
+		}
+
+		return preg_replace( '/[^\w-]/', '', $value );
+
+	}
+
+	/**
+	 * Setup an object with an array of data
+	 *
+	 * Overrides parent method to perform data validation
+	 * and sanitization, returning an error message when
+	 * validation issues are found.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $data key => val
+	 * @return LLMS_Form_Field_Data|WP_Error Field instance on success or error when validation errors are found.
+	 */
+	public function setup( $data ) {
+
+		$err = new WP_Error();
+		foreach ( $data as $key => &$value ) {
+
+			if ( ! self::validate( $key, $value ) ) {
+				$err->add(
+					sprintf( 'llms-form-field-invalid-data--%s', $key ),
+					sprintf( __( 'Invalid value submitted for %s', 'lifterlms' ), $key ),
+					compact( 'value' )
+				);
+			}
+
+			$value = self::sanitize( $key, $value );
+
+		}
+
+		if ( $err->has_errors() ) {
+			return $err;
+		}
+
+		return parent::setup( $data );
+
+	}
+
+	/**
+	 * Validate data by column
+	 *
+	 * @since [version]
+	 *
+	 * @param string $key   Column name.
+	 * @param mixed  $value Data to be inserted.
+	 * @return boolean
+	 */
+	public static function validate( $key, $value ) {
+
+		$ret = true;
+
+		switch ( $key ) {
+			case 'id':
+				$ret = preg_match( self::ID_FORMAT, $value );
+				break;
+			case 'name':
+			case 'store_key':
+				$ret = preg_match( '/[\w-]/', $value );
+				break;
+			case 'field_type':
+				$ret = in_array( $value, self::get_enum_field_types(), true );
+				break;
+			case 'store':
+				$ret = in_array( $value, self::get_enum_stores(), true );
+				break;
+		}
+
+		return $ret;
+
+	}
+
+
+
+
+
+
+
+	public function add_meta( $meta_key, $meta_value = '' ) {
+
+		$metas = $this->get_metas_obj( $meta_key );
+		return $metas->exists() ? false : $this->write_meta( $metas, $meta_key, $meta_value );
+
+	}
+
+	public function set_meta( $meta_key, $meta_value = '' ) {
+		return $this->write_meta( $this->get_metas_obj( $meta_key ), $meta_key, $meta_value );
+	}
+
+
+	public function get_meta( $meta_key = null ) {
+
+		if ( empty( $meta_key ) ) {
+			return $this->get_metas_obj()->get_all();
+		}
+
+		return $this->get_metas_obj( $meta_key )->get( 'meta_value' );
+
+	}
+
+	public function del_meta( $meta_key ) {
+		return $this->get_metas_obj( $meta_key )->delete();
+	}
+
+
+	private function get_metas_obj( $meta_key = null ) {
+		return new LLMS_Form_Field_Meta_Data( $this->get_id(), $meta_key );
+	}
```

Comment:
is there a reason why we instantiate a new `LLMS_Form_Field_Meta_Data` each time we need to refer to it?
Can't we just do it in this constructor and then use a class member?

Reply:
forget about the constructor stuff. I just misread.
So mmm we can just actually _cache_ this by `$meta_key`.